### PR TITLE
Submission of First Objective for Custom Abundance Widget Idea - GSOC

### PR DESCRIPTION
**Pull request by antreev-brar**
This is my attempt to solve the first objective of "Custom Abundance Widget" idea for GSOC 2021

## Description

The objective was to create a plot showing the abundance in each cell as a function of velocity using the tardis simulation set up using the csvy model format. To complete this objective I made the following changes :
- ```First_obj.ipynb``` -> Notebook containing the plot 
- ```first_obj_yml.yml``` -> .yml to run the run the tardis simulation
- ```first_obj_csvy.csvy ```-> csvy file that contains the custom model for this simulation with 6 shells for 4 elements (abundances have been assigned random values for now)
-  ```abundance_velocity_plot.py``` -> This file organises all plotting functions in an organised manner, this makes code clean and easy to understand. Inspired from SDECPlotter , AVData and AVPlotter classes have been introduced so that further features can be easily incorporated. Further functions implemented during the coding period can be introduced in similar manner.
-  ```visualisation/__init__.py ``` -> Called AVPlotter into ```__init__.py ``` so  that plotting functions can be called directly from notebook.
There wasn't any information regarding which velocity is supposed to be used therefore I created the plot by assuming it be inner shell boundary for each shell.
For the plot, there are various options like bar plot, line plot or scatter plot. For this objective, I preferred to use the line plot.

## Motivation and Context

I believe with some guidance I can achieve better results as in this plot, I have assumed a lot of parameters. Please let me know if there are any changes I should do. I'll try my best.

## How Has This Been Tested?
- [ ] Testing pipeline
- [ ] Reference Data Comparison following these [instructions](https://tardis-sn.github.io/tardis/development/continuous_integration.html)
- [ ] Other (please describe)

### Plot
Screenshot 2021-03-19 at 10 12 37 AM
Screenshot 2021-03-19 at 10 12 27 AM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] None of the above (please describe)

## Checklist:

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have built the documentation on my fork following [these instructions](https://tardis-sn.github.io/tardis/development/documentation_preview.html)
- [ ] I have assigned and requested two reviewers for this pull request

### CODE REVIEW

**Comment by jamesgillanders** [comment]
can you add velocity units to the x-axis?

**Comment by antreev-brar** [comment]
@andrewfullard @jamesgillanders @marxwillia I have updated the plots and also introduced plotting functions in ```visualisation/tools ``` , so that we can call the plotting functions directly to the notebook. 2 of my check in codecov are failing , can anyone guide me on how to resolve these issues and how to proceed further. Any help would be appreciated

**Comment by antreev-brar** [comment]
@wkerzendorf @jaladh-singhal  I have further organized the code into classes (inspired from SDEC plot) to integrate the matplotlib and plotly functions. Any new feature introduced during the coding period can now be easily incorporated into the flexible framework.

**Comment by andrewfullard** [comment]
I recommend beginning your application following the guidelines on the TARDIS GSoC website

**Review by jaladh-singhal — COMMENTED** [review]
@antreev-brar Sorry for such delayed reply. This looks great esp. the way you've structured AVPlotter & AVData classes. IMO a bit more work can be done to improve line plot aesthetics though.

You should start drafting your proposal now like Andrew said.

**Comment by jaladh-singhal** [comment]
Also the code formatting is not correct - I hope you can setup black correctly using: https://tardis-sn.github.io/tardis/development/code_quality.html#black and reformat it.

File: tardis/visualization/tools/abundance_velocity_plot.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+from tardis.util.base import atomic_number2element_symbol
+import plotly.graph_objects as go
+import numpy as np
+    
+
+class AVData:
+    """The data of simulation model used by Abundace Velocity (AV) Plot.
+
+    This preprocesses the data required by AVPlotter class for doing
+    calculations and plotting
```
Comment (marxwillia) [inline tardis/visualization/tools/abundance_velocity_plot.py:20]: Put the docstring in numpydoc format please.

File: tardis/visualization/tools/abundance_velocity_plot.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    def abundance_velocity_mpl(self):
+        """
+        Plots the abundance vs velocity of elements provided simulation result using matplotlib
+        
+        Returns
+        -------
+        None
+        """
+
+        plt.figure(figsize=(20,6))
```
Comment (marxwillia) [inline tardis/visualization/tools/abundance_velocity_plot.py:115]: Allow figsize to be set by user as argument.

File: tardis/visualization/tools/abundance_velocity_plot.py (hunk 1 above)
Comment (antreev-brar) [reply tardis/visualization/tools/abundance_velocity_plot.py:20]: Sure i am working on it

**Comment by antreev-brar** [comment]
@jaladh-singhal I am working on the changes proposed and also have started working on the draft proposal. I will send it to as soon as possible so that you can suggest changes.

**Comment by antreev-brar** [comment]
@jaladh-singhal I have reformatted the code. Now it passes all the tests.

**Pull request closed without merging**